File: Sources/YTools/Services/ClipboardHistoryStore.swift
import CryptoKit
import Foundation
import ImageIO

enum ClipboardStoreLoadResult: Sendable {
    case missing
    case loaded([ClipboardHistoryItem], warning: String?)
    case unavailable(String)
    case corrupted(String)
}

enum ClipboardStoreError: Error, LocalizedError, Sendable {
    case keyUnavailable(String)
    case readFailed(String)
    case corrupted(String)
    case writeFailed(String)

    var errorDescription: String? {
        switch self {
        case let .keyUnavailable(message), let .readFailed(message),
             let .corrupted(message), let .writeFailed(message): message
        }
    }
}

/// Incremental encrypted clipboard vault.
///
/// The small manifest is rewritten atomically, while immutable record payloads
/// and image thumbnails are encrypted independently. This avoids re-encoding
/// every historical image for each clipboard change.
final class ClipboardHistoryStore {
    private struct Manifest: Codable {
        var version = 2
        var entries: [Entry]
    }

    private struct Entry: Codable {
        let id: UUID
        let kind: ClipboardHistoryItem.Kind
        let displayText: String
        let createdAt: Date
        let sourceApplication: String?
        let contentHash: String
        let encryptedByteCount: Int
        var isPinned: Bool
    }

    private struct Record: Codable {
        let payload: [String]
        let imageData: Data?
    }

    private let keyAccessor = KeychainKeyAccessor(
        service: "com.ztools.native.clipboard-history",
        account: "encryption-key-v1",
        missingKeyMessage: "剪贴板密文存在，但钥匙串密钥缺失；存储保持只读。",
        randomGenerationMessage: "无法生成剪贴板加密密钥。"
    )
    private let legacyFileURL: URL
    private let vaultURL: URL
    private let manifestURL: URL
    private let recordsURL: URL
    private let thumbnailsURL: URL
    private let maximumEncryptedBytes = 200 * 1_024 * 1_024

    init(fileManager: FileManager = .default) {
        let applicationSupport = fileManager.urls(
            for: .applicationSupportDirectory,
            in: .userDomainMask
        ).first ?? fileManager.homeDirectoryForCurrentUser.appendingPathComponent("Library/Application Support")
        // Retain the original storage namespace so the product rename does not
        // orphan an existing encrypted clipboard archive.
        let root = applicationSupport.appendingPathComponent("ZToolsNative", isDirectory: true)
        legacyFileURL = root.appendingPathComponent("clipboard-history.v1.enc")
        vaultURL = root.appendingPathComponent("clipboard-vault-v2", isDirectory: true)
        manifestURL = vaultURL.appendingPathComponent("manifest.enc")
        recordsURL = vaultURL.appendingPathComponent("records", isDirectory: true)
        thumbnailsURL = vaultURL.appendingPathComponent("thumbnails", isDirectory: true)
        try? createVaultDirectories(fileManager: fileManager, root: root)
    }

    func load() -> ClipboardStoreLoadResult {
        if FileManager.default.fileExists(atPath: manifestURL.path) {
            return loadVersion2()
        }
        guard FileManager.default.fileExists(atPath: legacyFileURL.path) else { return .missing }
        switch loadLegacy() {
        case let .loaded(items, _):
            do {
                let migrated = try persist(items, originalImages: Dictionary(
                    uniqueKeysWithValues: items.compactMap { item in
                        item.kind == .image ? item.binaryData.map { (item.id, $0) } : nil
                    }
                ))
                let migratedURL = legacyFileURL.appendingPathExtension("migrated")
                try? FileManager.default.moveItem(at: legacyFileURL, to: migratedURL)
                return .loaded(migrated, warning: "剪贴板历史已迁移到增量加密存储。")
            } catch {
                return .unavailable("旧剪贴板历史读取成功，但迁移失败：\(error.localizedDescription)")
            }
        case .missing:
            return .missing
        case let .unavailable(message):
            return .unavailable(message)
        case let .corrupted(message):
            return .corrupted(message)
        }
    }

    /// Persists order/metadata and writes payload files only for new records.
    /// The returned items contain thumbnails rather than full image data.
    func persist(
        _ requestedItems: [ClipboardHistoryItem],
        originalImages: [UUID: Data] = [:]
    ) throws -> [ClipboardHistoryItem] {
        let key = try encryptionKey(createIfMissing: true)
        let existing = try currentManifest(key: key)
        let existingByID = Dictionary(uniqueKeysWithValues: existing.entries.map { ($0.id, $0) })
        var candidateEntries: [Entry] = []
        var normalizedByID: [UUID: ClipboardHistoryItem] = [:]
        var newlyWrittenIDs: Set<UUID> = []

        do {
            for item in requestedItems {
                if let old = existingByID[item.id] {
                    candidateEntries.append(Entry(
                        id: old.id,
                        kind: old.kind,
                        displayText: item.displayText,
                        createdAt: old.createdAt,
                        sourceApplication: item.sourceApplication,
                        contentHash: item.contentHash ?? old.contentHash,
                        encryptedByteCount: old.encryptedByteCount,
                        isPinned: item.pinned
                    ))
                    normalizedByID[item.id] = item
                    continue
                }

                let originalImage = item.kind == .image ? (originalImages[item.id] ?? item.binaryData) : nil
                let record = Record(payload: item.payload, imageData: originalImage)
                let clear = try JSONEncoder().encode(record)
                let encrypted = try seal(clear, key: key)
                try writeProtected(encrypted, to: recordURL(item.id))
                newlyWrittenIDs.insert(item.id)

                var normalized = item
                if item.kind == .image, let originalImage {
                    let thumbnail = makeThumbnail(from: originalImage) ?? Data()
                    if !thumbnail.isEmpty {
                        try writeProtected(try seal(thumbnail, key: key), to: thumbnailURL(item.id))
                    }
                    normalized = ClipboardHistoryItem(
                        id: item.id,
                        kind: item.kind,
                        payload: item.payload,
                        createdAt: item.createdAt,
                        sourceApplication: item.sourceApplication,
                        binaryData: thumbnail.isEmpty ? nil : thumbnail,
                        contentHash: item.contentHash ?? hash(clear),
                        isPinned: item.pinned
                    )
                }
                normalizedByID[item.id] = normalized
                candidateEntries.append(Entry(
                    id: item.id,
                    kind: item.kind,
                    displayText: item.displayText,
                    createdAt: item.createdAt,
                    sourceApplication: item.sourceApplication,
                    contentHash: item.contentHash ?? hash(clear),
                    encryptedByteCount: encrypted.count,
                    isPinned: item.pinned
                ))
            }

            var retainedEntries: [Entry] = []
            var totalBytes = 0
            for entry in candidateEntries {
                guard totalBytes + entry.encryptedByteCount <= maximumEncryptedBytes else { continue }
                retainedEntries.append(entry)
                totalBytes += entry.encryptedByteCount
            }
            let retainedIDs = Set(retainedEntries.map(\.id))
            let manifest = Manifest(entries: retainedEntries)
            let manifestData = try JSONEncoder().encode(manifest)
            try writeProtected(try seal(manifestData, key: key), to: manifestURL)

            let obsoleteIDs = Set(existing.entries.map(\.id)).union(newlyWrittenIDs).subtracting(retainedIDs)
            obsoleteIDs.forEach(removePayloadFiles)
            return retainedEntries.compactMap { normalizedByID[$0.id] }
        } catch {
            newlyWrittenIDs.forEach(removePayloadFiles)
            throw error
        }
    }

    func imageData(for id: UUID) -> Data? {
        guard let key = try? encryptionKey(createIfMissing: false),
              let encrypted = try? Data(contentsOf: recordURL(id)),
              let clear = try? open(encrypted, key: key),
              let record = try? JSONDecoder().decode(Record.self, from: clear) else { return nil }
        return record.imageData
    }

    func removePersistedHistory() throws {
        if FileManager.default.fileExists(atPath: vaultURL.path) {
            try FileManager.default.removeItem(at: vaultURL)
        }
        if FileManager.default.fileExists(atPath: legacyFileURL.path) {
            try FileManager.default.removeItem(at: legacyFileURL)
        }
        let migratedLegacy = legacyFileURL.appendingPathExtension("migrated")
        if FileManager.default.fileExists(atPath: migratedLegacy.path) {
            try FileManager.default.removeItem(at: migratedLegacy)
        }
        let root = vaultURL.deletingLastPathComponent()
        try createVaultDirectories(fileManager: .default, root: root)
    }

    func diskUsage() -> Int64 {
        guard let enumerator = FileManager.default.enumerator(
            at: vaultURL,
            includingPropertiesForKeys: [.fileSizeKey, .isRegularFileKey],
            options: [.skipsHiddenFiles]
        ) else { return 0 }
        var total: Int64 = 0
        for case let url as URL in enumerator {
            guard let values = try? url.resourceValues(forKeys: [.fileSizeKey, .isRegularFileKey]),
                  values.isRegularFile == true else { continue }
            total += Int64(values.fileSize ?? 0)
        }
        return total
    }

    private func loadVersion2() -> ClipboardStoreLoadResult {
        let key: Data
        do {
            key = try encryptionKey(createIfMissing: false)
        } catch {
            return .unavailable(error.localizedDescription)
        }
        let manifest: Manifest
        do {
            let encrypted = try Data(contentsOf: manifestURL)
            let clear = try open(encrypted, key: key)
            manifest = try JSONDecoder().decode(Manifest.self, from: clear)
            guard manifest.version == 2 else {
                return .corrupted("不支持的剪贴板存储版本：\(manifest.version)")
            }
        } catch {
            return .corrupted("剪贴板清单验证失败；原文件未被覆盖：\(error.localizedDescription)")
        }

        var items: [ClipboardHistoryItem] = []
        var skipped = 0
        for entry in manifest.entries {
            do {
                let payload: [String]
                let thumbnail: Data?
                switch entry.kind {
                case .image:
                    payload = [entry.displayText]
                    thumbnail = try loadThumbnail(entry.id, key: key)
                case .text, .files:
                    let encrypted = try Data(contentsOf: recordURL(entry.id))
                    let clear = try open(encrypted, key: key)
                    payload = try JSONDecoder().decode(Record.self, from: clear).payload
                    thumbnail = nil
                }
                items.append(ClipboardHistoryItem(
                    id: entry.id,
                    kind: entry.kind,
                    payload: payload,
                    createdAt: entry.createdAt,
                    sourceApplication: entry.sourceApplication,
                    binaryData: thumbnail,
                    contentHash: entry.contentHash,
                    isPinned: entry.isPinned
                ))
            } catch {
                skipped += 1
            }
        }
        let warning = skipped == 0 ? nil : "有 \(skipped) 条剪贴板记录损坏，已跳过但未覆盖原密文。"
        return .loaded(items, warning: warning)
    }

    private func loadLegacy() -> ClipboardStoreLoadResult {
        let key: Data
        do {
            key = try encryptionKey(createIfMissing: false)
        } catch {
            return .unavailable(error.localizedDescription)
        }
        do {
            let encrypted = try Data(contentsOf: legacyFileURL)
            let clear = try open(encrypted, key: key)
            let items = try JSONDecoder().decode([ClipboardHistoryItem].self, from: clear)
            return .loaded(items.map(addingHashIfNeeded), warning: nil)
        } catch {
            return .corrupted("旧剪贴板历史验证失败；原文件未被覆盖：\(error.localizedDescription)")
        }
    }

    private func currentManifest(key: Data) throws -> Manifest {
        guard FileManager.default.fileExists(atPath: manifestURL.path) else { return Manifest(entries: []) }
        let clear = try open(Data(contentsOf: manifestURL), key: key)
        return try JSONDecoder().decode(Manifest.self, from: clear)
    }

    private func addingHashIfNeeded(_ item: ClipboardHistoryItem) -> ClipboardHistoryItem {
        guard item.contentHash == nil else { return item }
        let data: Data
        switch item.kind {
        case .text, .files: data = Data(item.payload.joined(separator: "\0").utf8)
        case .image: data = item.binaryData ?? Data()
        }
        return ClipboardHistoryItem(
            id: item.id,
            kind: item.kind,
            payload: item.payload,
            createdAt: item.createdAt,
            sourceApplication: item.sourceApplication,
            binaryData: item.binaryData,
            contentHash: hash(data),
            isPinned: item.pinned
        )
    }

    private func encryptionKey(createIfMissing: Bool) throws -> Data {
        do {
            return try keyAccessor.key(createIfMissing: createIfMissing)
        } catch {
            throw ClipboardStoreError.keyUnavailable(error.localizedDescription)
        }
    }

    private func seal(_ clear: Data, key: Data) throws -> Data {
        guard let combined = try AES.GCM.seal(clear, using: SymmetricKey(data: key)).combined else {
            throw ClipboardStoreError.writeFailed("AES-GCM 未生成组合密文。")
        }
        return combined
    }

    private func open(_ encrypted: Data, key: Data) throws -> Data {
        do {
            return try AES.GCM.open(
                AES.GCM.SealedBox(combined: encrypted),
                using: SymmetricKey(data: key)
            )
        } catch {
            throw ClipboardStoreError.corrupted("AES-GCM 验证失败：\(error.localizedDescription)")
        }
    }

    private func writeProtected(_ data: Data, to url: URL) throws {
        do {
            try data.write(to: url, options: .atomic)
            try FileManager.default.setAttributes([.posixPermissions: 0o600], ofItemAtPath: url.path)
        } catch {
            throw ClipboardStoreError.writeFailed("无法写入 \(url.lastPathComponent)：\(error.localizedDescription)")
        }
    }

    private func createVaultDirectories(fileManager: FileManager, root: URL) throws {
        for directory in [root, vaultURL, recordsURL, thumbnailsURL] {
            try fileManager.createDirectory(at: directory, withIntermediateDirectories: true)
            try? fileManager.setAttributes([.posixPermissions: 0o700], ofItemAtPath: directory.path)
        }
    }

    private func recordURL(_ id: UUID) -> URL { recordsURL.appendingPathComponent("\(id.uuidString).enc") }
    private func thumbnailURL(_ id: UUID) -> URL { thumbnailsURL.appendingPathComponent("\(id.uuidString).enc") }

    private func removePayloadFiles(_ id: UUID) {
        try? FileManager.default.removeItem(at: recordURL(id))
        try? FileManager.default.removeItem(at: thumbnailURL(id))
    }

    private func loadThumbnail(_ id: UUID, key: Data) throws -> Data? {
        let url = thumbnailURL(id)
        guard FileManager.default.fileExists(atPath: url.path) else { return nil }
        return try open(Data(contentsOf: url), key: key)
    }

    private func makeThumbnail(from data: Data) -> Data? {
        guard let source = CGImageSourceCreateWithData(data as CFData, nil),
              let image = CGImageSourceCreateThumbnailAtIndex(source, 0, [
                  kCGImageSourceCreateThumbnailFromImageAlways: true,
                  kCGImageSourceCreateThumbnailWithTransform: true,
                  kCGImageSourceThumbnailMaxPixelSize: 192
              ] as CFDictionary) else { return nil }
        let output = NSMutableData()
        guard let destination = CGImageDestinationCreateWithData(
            output,
            "public.png" as CFString,
            1,
            nil
        ) else { return nil }
        CGImageDestinationAddImage(destination, image, nil)
        return CGImageDestinationFinalize(destination) ? output as Data : nil
    }

    private func hash(_ data: Data) -> String {
        SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
    }

}

File: src/YTools.Windows/Services/Storage/ClipboardHistoryStore.cs
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Drawing.Imaging;
using System.IO;
using System.Security.Cryptography;
using System.Text.Json;
using YTools.Infrastructure;
using YTools.Models;

namespace YTools.Services.Storage;

public enum ClipboardStoreLoadResultKind
{
    Missing,
    Loaded,
    Unavailable,
    Corrupted
}

public sealed record ClipboardStoreLoadResult(
    ClipboardStoreLoadResultKind Kind,
    IReadOnlyList<ClipboardHistoryItem>? Items,
    string? Warning)
{
    public static ClipboardStoreLoadResult Missing() =>
        new(ClipboardStoreLoadResultKind.Missing, null, null);

    public static ClipboardStoreLoadResult Loaded(IReadOnlyList<ClipboardHistoryItem> items, string? warning) =>
        new(ClipboardStoreLoadResultKind.Loaded, items, warning);

    public static ClipboardStoreLoadResult Unavailable(string message) =>
        new(ClipboardStoreLoadResultKind.Unavailable, null, message);

    public static ClipboardStoreLoadResult Corrupted(string message) =>
        new(ClipboardStoreLoadResultKind.Corrupted, null, message);
}

/// <summary>
/// Incremental encrypted clipboard vault: a small manifest plus independently
/// encrypted immutable record payloads and image thumbnails.
/// </summary>
public sealed class ClipboardHistoryStore
{
    private const int MaximumEncryptedBytes = 200 * 1024 * 1024;

    private readonly string _vaultDirectory;
    private readonly string _manifestFile;
    private readonly string _recordsDirectory;
    private readonly string _thumbnailsDirectory;

    public ClipboardHistoryStore()
    {
        AppPaths.EnsureDirectories();
        _vaultDirectory = Path.Combine(AppPaths.VaultDirectory, "clipboard-vault-v2");
        _manifestFile = Path.Combine(_vaultDirectory, "manifest.enc");
        _recordsDirectory = Path.Combine(_vaultDirectory, "records");
        _thumbnailsDirectory = Path.Combine(_vaultDirectory, "thumbnails");
        Directory.CreateDirectory(_vaultDirectory);
        Directory.CreateDirectory(_recordsDirectory);
        Directory.CreateDirectory(_thumbnailsDirectory);
    }

    public ClipboardStoreLoadResult Load()
    {
        if (!File.Exists(_manifestFile))
        {
            return ClipboardStoreLoadResult.Missing();
        }

        byte[] key;
        try
        {
            key = DpapiKeyAccessor.Key(createIfMissing: false);
        }
        catch (Exception exception)
        {
            return ClipboardStoreLoadResult.Unavailable(exception.Message);
        }

        Manifest manifest;
        try
        {
            var encrypted = File.ReadAllBytes(_manifestFile);
            var clear = AesGcmBox.Open(encrypted, key);
            manifest = JsonSerializer.Deserialize<Manifest>(clear)
                ?? throw new InvalidDataException("清单为空");
            if (manifest.Version != 2)
            {
                return ClipboardStoreLoadResult.Corrupted(
                    $"不支持的剪贴板存储版本：{manifest.Version}");
            }
        }
        catch (Exception exception)
        {
            return ClipboardStoreLoadResult.Corrupted(
                $"剪贴板清单验证失败；原文件未被覆盖：{exception.Message}");
        }

        var items = new List<ClipboardHistoryItem>();
        var skipped = 0;
        foreach (var entry in manifest.Entries)
        {
            try
            {
                IReadOnlyList<string> payload;
                byte[]? thumbnail = null;
                switch (entry.Kind)
                {
                    case ClipboardItemKind.Image:
                        payload = [entry.DisplayText];
                        thumbnail = LoadThumbnail(entry.Id, key);
                        break;
                    case ClipboardItemKind.Text:
                    case ClipboardItemKind.Files:
                        var encrypted = File.ReadAllBytes(RecordFile(entry.Id));
                        var clear = AesGcmBox.Open(encrypted, key);
                        var record = JsonSerializer.Deserialize<Record>(clear)
                            ?? throw new InvalidDataException("记录为空");
                        payload = record.Payload;
                        break;
                    default:
                        throw new InvalidDataException("未知记录类型");
                }

                items.Add(new ClipboardHistoryItem(
                    entry.Id,
                    entry.Kind,
                    payload,
                    entry.CreatedAt,
                    entry.SourceApplication,
                    thumbnail,
                    entry.ContentHash,
                    entry.IsPinned,
                    entry.UpdatedAt,
                    Math.Max(1, entry.CopyCount)));
            }
            catch
            {
                skipped += 1;
            }
        }

        var warning = skipped == 0 ? null : $"有 {skipped} 条剪贴板记录损坏，已跳过但未覆盖原密文。";
        return ClipboardStoreLoadResult.Loaded(items, warning);
    }

    public List<ClipboardHistoryItem> Persist(
        IReadOnlyList<ClipboardHistoryItem> requestedItems,
        IReadOnlyDictionary<Guid, byte[]> originalImages)
    {
        var key = DpapiKeyAccessor.Key(createIfMissing: true);
        var existing = CurrentManifest(key);
        var existingById = existing.Entries.ToDictionary(entry => entry.Id);
        var candidateEntries = new List<Entry>();
        var normalizedById = new Dictionary<Guid, ClipboardHistoryItem>();
        var newlyWrittenIds = new HashSet<Guid>();

        try
        {
            foreach (var item in requestedItems)
            {
                if (existingById.TryGetValue(item.Id, out var old))
                {
                    candidateEntries.Add(new Entry(
                        old.Id,
                        old.Kind,
                        item.DisplayText,
                        old.CreatedAt,
                        item.SourceApplication,
                        item.ContentHash ?? old.ContentHash,
                        old.EncryptedByteCount,
                        item.IsPinned,
                        item.UpdatedAt ?? old.UpdatedAt,
                        Math.Max(1, item.CopyCount)));
                    normalizedById[item.Id] = item;
                    continue;
                }

                var originalImage = item.Kind == ClipboardItemKind.Image
                    ? (originalImages.TryGetValue(item.Id, out var image) ? image : item.BinaryData)
                    : null;
                var record = new Record(item.Payload.ToList(), originalImage);
                var clear = JsonSerializer.SerializeToUtf8Bytes(record);
                var encrypted = AesGcmBox.Seal(clear, key);
                WriteProtected(encrypted, RecordFile(item.Id));
                newlyWrittenIds.Add(item.Id);

                var normalized = item;
                byte[]? thumbnail = null;
                if (item.Kind == ClipboardItemKind.Image && originalImage is not null)
                {
                    thumbnail = MakeThumbnail(originalImage);
                    if (thumbnail is { Length: > 0 })
                    {
                        WriteProtected(AesGcmBox.Seal(thumbnail, key), ThumbnailFile(item.Id));
                    }

                    normalized = item with { BinaryData = thumbnail is { Length: > 0 } ? thumbnail : null };
                }

                normalizedById[item.Id] = normalized;
                candidateEntries.Add(new Entry(
                    item.Id,
                    item.Kind,
                    item.DisplayText,
                    item.CreatedAt,
                    item.SourceApplication,
                    item.ContentHash ?? Hash(clear),
                    encrypted.Length,
                    item.IsPinned,
                    item.UpdatedAt,
                    Math.Max(1, item.CopyCount)));
            }

            var retainedEntries = new List<Entry>();
            var totalBytes = 0L;
            foreach (var entry in candidateEntries)
            {
                if (totalBytes + entry.EncryptedByteCount > MaximumEncryptedBytes)
                {
                    continue;
                }

                retainedEntries.Add(entry);
                totalBytes += entry.EncryptedByteCount;
            }

            var retainedIds = retainedEntries.Select(entry => entry.Id).ToHashSet();
            var manifest = new Manifest { Entries = retainedEntries };
            var manifestData = JsonSerializer.SerializeToUtf8Bytes(manifest);
            WriteProtected(AesGcmBox.Seal(manifestData, key), _manifestFile);

            var obsoleteIds = existing.Entries.Select(entry => entry.Id)
                .Union(newlyWrittenIds)
                .Where(id => !retainedIds.Contains(id))
                .ToList();
            foreach (var id in obsoleteIds)
            {
                RemovePayloadFiles(id);
            }

            return retainedEntries
                .Select(entry => normalizedById[entry.Id])
                .ToList();
        }
        catch
        {
            foreach (var id in newlyWrittenIds)
            {
                RemovePayloadFiles(id);
            }

            throw;
        }
    }

    public byte[]? ImageData(Guid id)
    {
        try
        {
            var key = DpapiKeyAccessor.Key(createIfMissing: false);
            var encrypted = File.ReadAllBytes(RecordFile(id));
            var clear = AesGcmBox.Open(encrypted, key);
            var record = JsonSerializer.Deserialize<Record>(clear);
            return record?.ImageData;
        }
        catch
        {
            return null;
        }
    }

    public void RemovePersistedHistory()
    {
        if (Directory.Exists(_vaultDirectory))
        {
            Directory.Delete(_vaultDirectory, recursive: true);
        }

        Directory.CreateDirectory(_vaultDirectory);
        Directory.CreateDirectory(_recordsDirectory);
        Directory.CreateDirectory(_thumbnailsDirectory);
    }

    public long DiskUsage()
    {
        if (!Directory.Exists(_vaultDirectory))
        {
            return 0;
        }

        try
        {
            return Directory.EnumerateFiles(_vaultDirectory, "*", SearchOption.AllDirectories)
                .Sum(path => new FileInfo(path).Length);
        }
        catch
        {
            return 0;
        }
    }

    private Manifest CurrentManifest(byte[] key)
    {
        if (!File.Exists(_manifestFile))
        {
            return new Manifest();
        }

        var clear = AesGcmBox.Open(File.ReadAllBytes(_manifestFile), key);
        return JsonSerializer.Deserialize<Manifest>(clear) ?? new Manifest();
    }

    private byte[]? LoadThumbnail(Guid id, byte[] key)
    {
        var file = ThumbnailFile(id);
        if (!File.Exists(file))
        {
            return null;
        }

        return AesGcmBox.Open(File.ReadAllBytes(file), key);
    }

    private void WriteProtected(byte[] data, string path)
    {
        File.WriteAllBytes(path, data);
        AppPaths.RestrictFile(path);
    }

    private string RecordFile(Guid id) => Path.Combine(_recordsDirectory, $"{id:N}.enc");

    private string ThumbnailFile(Guid id) => Path.Combine(_thumbnailsDirectory, $"{id:N}.enc");

    private void RemovePayloadFiles(Guid id)
    {
        try
        {
            File.Delete(RecordFile(id));
        }
        catch
        {
            // Best-effort cleanup.
        }

        try
        {
            File.Delete(ThumbnailFile(id));
        }
        catch
        {
            // Best-effort cleanup.
        }
    }

    private static byte[]? MakeThumbnail(byte[] data)
    {
        try
        {
            using var stream = new MemoryStream(data);
            using var image = Image.FromStream(stream);
            var maxSide = 192;
            var scale = Math.Min(1.0, maxSide / (double)Math.Max(image.Width, image.Height));
            var width = Math.Max(1, (int)(image.Width * scale));
            var height = Math.Max(1, (int)(image.Height * scale));
            using var thumbnail = new Bitmap(width, height);
            using (var graphics = Graphics.FromImage(thumbnail))
            {
                graphics.InterpolationMode = InterpolationMode.HighQualityBicubic;
                graphics.SmoothingMode = SmoothingMode.HighQuality;
                graphics.DrawImage(image, 0, 0, width, height);
            }

            using var output = new MemoryStream();
            thumbnail.Save(output, ImageFormat.Png);
            return output.ToArray();
        }
        catch
        {
            return null;
        }
    }

    private static string Hash(byte[] data)
    {
        return Convert.ToHexString(SHA256.HashData(data)).ToLowerInvariant();
    }

    private sealed class Manifest
    {
        public int Version { get; set; } = 2;

        public List<Entry> Entries { get; set; } = [];
    }

    private sealed class Entry
    {
        public Entry()
        {
        }

        public Entry(
            Guid id,
            ClipboardItemKind kind,
            string displayText,
            DateTimeOffset createdAt,
            string? sourceApplication,
            string contentHash,
            int encryptedByteCount,
            bool isPinned,
            DateTimeOffset? updatedAt,
            int copyCount)
        {
            Id = id;
            Kind = kind;
            DisplayText = displayText;
            CreatedAt = createdAt;
            SourceApplication = sourceApplication;
            ContentHash = contentHash;
            EncryptedByteCount = encryptedByteCount;
            IsPinned = isPinned;
            UpdatedAt = updatedAt;
            CopyCount = copyCount;
        }

        public Guid Id { get; set; }

        public ClipboardItemKind Kind { get; set; }

        public string DisplayText { get; set; } = "";

        public DateTimeOffset CreatedAt { get; set; }

        public string? SourceApplication { get; set; }

        public string ContentHash { get; set; } = "";

        public int EncryptedByteCount { get; set; }

        public bool IsPinned { get; set; }

        public DateTimeOffset? UpdatedAt { get; set; }

        public int CopyCount { get; set; } = 1;
    }

    private sealed class Record
    {
        public Record()
        {
        }

        public Record(List<string> payload, byte[]? imageData)
        {
            Payload = payload;
            ImageData = imageData;
        }

        public List<string> Payload { get; set; } = [];

        public byte[]? ImageData { get; set; }
    }
}
